### PR TITLE
onedrive: minor opt of quickxorhash

### PR DESCRIPTION
**Pull request by isaacl**
Avoid copying the hash array.

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-new-feature-or-bug-fix).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: backend/onedrive/quickxorhash/quickxorhash.go [hunk 1]
```diff
@@ -136,7 +136,8 @@ func (q *quickXorHash) Write(p []byte) (n int, err error) {
 func (q *quickXorHash) checkSum() (h [Size]byte) {
 	// Output the data as little endian bytes
 	ph := 0
-	for _, d := range q.data[:len(q.data)-1] {
```
Comment (ncw) [inline backend/onedrive/quickxorhash/quickxorhash.go:139]: This doesn't copy the `q.data` array, it slices it, so the backing store remains in the same place. Thus its quite a cheap thing to do as all it is is filling in 3 fields in the slice header.

If you look at a disassembly you'll see that your code saves about 5 instructions of setup.

https://godbolt.org/z/jh1rM7coY

So it is probably worth doing but the difference it makes will be very small.

**Review by ncw — COMMENTED** [review]
I don't think this makes as much performance impact as you might think - see comments inline! It is probably worth merging though.

Thoughts?

**Comment by ncw** [comment]
> ok! I had read on go documentation that slices copy the backing array. Plus of course, the main expense of this hashing algorithm is in the hashing part, not the digest. I’m not sure if there’s any way to make this hasher faster, it’s a complicated af algorithm

I haven't made any attempt to optimize this so I expect there is low hanging fruit!

If you want to have a go I'd make a benchmark, maybe something like this one

https://cs.opensource.google/go/go/+/refs/tags/go1.17.6:src/crypto/md5/md5_test.go;l=218

then try to eliminate all array bounds checks from the loop here (article https://go101.org/article/bounds-check-elimination.html )

https://github.com/rclone/rclone/blob/9d4eab32d82c304666d7d2331ab3c4f6ca0ed3eb/backend/onedrive/quickxorhash/quickxorhash.go#L72

I'd be surprised if you couldn't get it to run 10 or 20% faster with only a few small changes.

I'll merge what you've done as it makes a difference - thank you.

**Comment by isaacl** [comment]
How do you feel about the using pointers (`package unsafe`)? (btw I can't join slack due to #5887)

**Comment by ncw** [comment]
> How do you feel about the using pointers (`package unsafe`)?

I'd prefer not to use unsafe. I don't think it will make stuff go a lot faster... though I could be proved wrong by benchmarks!

**Pull request merged**